in_toto/gpg/common.py

```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
-    construct a public key dictionary, containing a master key and optional
-    subkeys, where either the master key or the subkeys are identified by
-    the passed keyid.
-
-    NOTE: If the keyid matches one of the subkeys, a warning is issued to
-    notify the user about potential privilege escalation.
+    Parse and verify passed gpg public key data and return the master
+    key (aka. primary key) including certified information (e.g. key
+    expiration date), and corresponding subkeys bound to the primary key via
+    signatures.
 
   <Arguments>
     data:
           Public key data as written to stdout by GPG_EXPORT_PUBKEY_COMMAND.
 
   <Exceptions>
-    ValueError:
-          If no data is passed
+    in_toto.gpg.exceptions.PacketParsingError
+          If data is empty.
+          If data cannot be parsed.
 
-    in_toto.gpg.exceptions.KeyNotFoundError
-          If neither the master key or one of the subkeys match the passed
-          keyid.
 
   <Side Effects>
     None.
 
   <Returns>
-    A public key in the format in_toto.gpg.formats.PUBKEY_SCHEMA containing
-    available subkeys, where either the master key or one of the subkeys match
-    the passed keyid.
+    A tuple of a public key in the format in_toto.gpg.formats.PUBKEY_SCHEMA
+    that contains the master key, and a list of public keys in the same format.
 
   """
-  master_public_key = None
-  sub_public_keys = {}
+  if not data:
+    raise PacketParsingError("Cannot parse keys from empty gpg data.")
+
+  # Temporary data structure to hold parsed gpg packets
+  key_bundle = {
+    PACKET_TYPE_PRIMARY_KEY: {
+      "key": None,
+      "packet": None,
+      "signatures": []
+    },
+    PACKET_TYPE_USER_ID: collections.OrderedDict(),
+    PACKET_TYPE_USER_ATTR: collections.OrderedDict(),
+    PACKET_TYPE_SUB_KEY: collections.OrderedDict()
+  }
+
+  # Iterate over gpg data and parse out packets of different types
+  position = 0
+  while position < len(data):
+    try:
+      payload, packet_length, packet_type = \
+          in_toto.gpg.util.parse_packet_header(data[position:])
+      body_len = len(payload)
+      packet = data[position:position+packet_length]
+
+      # The first (and only the first) packet in the bundle must be the master
+      # key.  See RFC4880 12.1 Key Structures, V4 version keys
+      # TODO: Do we need additional key structure assertions? e.g.
+      # - there must be least one User ID packet, or
+      # - order and type of signatures, or
+      # - disallow duplicate packets
+      if packet_type != PACKET_TYPE_PRIMARY_KEY and \
+          not key_bundle[PACKET_TYPE_PRIMARY_KEY]["key"]:
+        raise Exception("First packet must be a primary key ('{}'), got '{}'."
+            .format(PACKET_TYPE_PRIMARY_KEY, packet_type))
+
+      elif packet_type == PACKET_TYPE_PRIMARY_KEY and \
+          key_bundle[PACKET_TYPE_PRIMARY_KEY]["key"]:
+        raise Exception("Unexpected primary key.")
+
```

Comment:
I always find `if` cascades with complex causes difficult to read, as I'm never sure if I'm not missing a combination of conditions that's not handled.

The `elif` in L198-199 could be rolled into L205+ as a sub-`if` clause.

Reply:
Agreed, that's why I added comprehensive comments.

> The `elif` in L198-199 could be rolled into L205+ as a sub-`if` clause.
That's how I wrote it initially, but then I thought it would be more readable the way it is now, as it is less arrow-head'ish.

Reply:
> arrow-head'ish

Hmm, like, *two* levels in doesn't feel *that* spiky to me :-)

Reply:
It would be 4 levels already: `while`, `try`, `elif`, `if` ... 5 if you count the `def`... :P

Reply:
...which is *technically* correct, but I think the `def` and `while` and `try` only count as overarching brackets. So: two ;-)

Reply:
Agreed, it's only two (...technically). I'll still keep this one.